### PR TITLE
AK: Make StringUtils to_int a bit faster

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
#### AK: Add a max_digits field to NumericLimits

This field describes the maximum amount of digits a number of that type
can have in the decimal system.
The fromular used is:
```py
len(str(2**n -1))
```

#### LibLine: Use to_int instead of to_uint for time_t's

time_t is technically signed and this will allow us to make to_int and
to_uint more restrictive

#### AK: Do range checks earlier in to_[u]int

This avoids use of __builtins and speeds up the code by up to 10%.
This also adds a more turned down version of the benchmark used to
measure the speedup.

### CODE REVIEW

File: AK/StringUtils.cpp [hunk 1]
```diff
@@ -102,16 +103,19 @@ Optional<T> convert_to_int(StringView str, TrimWhitespace trim_whitespace)
             sign = -1;
     }
 
+    while (characters[i] == '0' && i < string.length())
+        ++i;
+
+    if ((string.length() - i) > NumericLimits<T>::max_digits())
+        return {};
```
Comment (davidot) [inline AK/StringUtils.cpp:110]: Just checking this does not mean we avoid overflow.
We can probably use a `max_digits_no_overflow` which is lower than your current `max_digits()`.

For example `9223372036854776808` has 19 digits but is higher than $2^{63}$.

This is a problem for every single integer type as powers of ten don't match the powers of two

File: AK/StringUtils.cpp [hunk 2]
```diff
@@ -132,17 +136,21 @@ Optional<T> convert_to_uint(StringView str, TrimWhitespace trim_whitespace)
...
-        if (characters[i] < '0' || characters[i] > '9')
-            return {};
+    size_t i = 0;
+    while (characters[i] == '0' && i < string.length())
+        ++i;
 
-        if (__builtin_mul_overflow(value, 10, &value))
-            return {};
+    if ((string.length() - i) > NumericLimits<T>::max_digits())
+        return {};
```
Comment (davidot) [inline AK/StringUtils.cpp:146]: Here an example would be `19000000000000000000` which is 20 digits but larger than $2^{64}$

**Review by davidot — CHANGES_REQUESTED** [review]
These methods are now no longer overflow safe and will give wrong results

**Comment by Hendiadyoin1** [comment]
I think I have a solution, although it did cost a surprising amount of runtime....

**Review by davidot — APPROVED** [review]
Looks fine to me now, yeah it seems tricky to do this safely unfortunately.
Maybe we can add tests to ensure this keeps working?

**Comment by Hendiadyoin1** [comment]
As far as I could tell all of CI caught it

**Comment by davidot** [comment]
> As far as I could tell all of CI caught it

Ah sorry I didn't check the CI before, good to know!

File: AK/NumericLimits.h [hunk 3]
```diff
@@ -26,76 +26,87 @@ struct NumericLimits<signed char> {
     static constexpr signed char min() { return -__SCHAR_MAX__ - 1; }
     static constexpr signed char max() { return __SCHAR_MAX__; }
     static constexpr bool is_signed() { return true; }
+    static constexpr size_t max_digits() { return 3; }
```
Comment (ADKaster) [inline AK/NumericLimits.h:29]: :de:  is showing, should be "formula" in the first commit message.

File: AK/StringUtils.cpp [hunk 4]
```diff
@@ -102,17 +104,35 @@ Optional<T> convert_to_int(StringView str, TrimWhitespace trim_whitespace)
...
             return {};
 
-        if (__builtin_mul_overflow(value, 10, &value))
-            return {};
+        value *= 10;
+        value += sign * (characters[i] - '0');
+    }
 
-        if (__builtin_add_overflow(value, sign * (characters[i] - '0'), &value))
+    // If we are at the edge of the representable range, we need to check for overflows
```
Comment (ADKaster) [inline AK/StringUtils.cpp:124]: is the 10% claim true across the board? or only for numbers close to NumericLimits::<T>::max() ?

In one of your other comment replies you mentioned that the may_overflow change cost "significant runtime", is it still a net positive of ~10%?

The benchmark looks like it only verifies large numbers, is there any noticable performance difference for small numbers? < 10, 10 - 100, < 1000? I feel like numbers in that range are more common in our codebase and the types of applications/data they commonly interact with.

**Review by ADKaster — COMMENTED** [review]
This looks like a reasonable refactor, but i've got a couple questions

**Pull request closed without merging**